truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedTruffleRuntime.java

```diff
@@ -912,10 +913,45 @@ private void notifyCompilationFailure(OptimizedCallTarget callTarget, Throwable
     protected void onEngineCreated(EngineData engine) {
     }
 
+    private long stoppedCompilationTime = 0;
+    private boolean logShutdownCompilations = true;
+
     @SuppressWarnings("try")
     public CompilationTask submitForCompilation(OptimizedCallTarget optimizedCallTarget, boolean lastTierCompilation) {
-        Priority priority = new Priority(optimizedCallTarget.getCallAndLoopCount(), lastTierCompilation ? Priority.Tier.LAST : Priority.Tier.FIRST);
-        return getCompileQueue().submitCompilation(priority, optimizedCallTarget);
+        BackgroundCompileQueue queue = getCompileQueue();
+        CompilationActivityMode compilationActivityMode = getCompilationActivityMode();
+        if (compilationActivityMode == CompilationActivityMode.RUN_COMPILATION ||
```

Comment:
Minor, but I think we could write this as a switch and easier to read retry handling (probably opinionated):

```
CompilationActivityMode compilationActivityMode = getCompilationActivityMode();
if (compilationActivityMode != CompilationActivityMode.RUN_COMPILATION) {
   if ((stoppedCompilationTime != 0 && System.currentTimeMillis() - stoppedCompilationTime > optimizedCallTarget.getOptionValue(StoppedCompilationRetryDelay)) {
        compilationActivityMode = CompilationActivityMode.RUN_COMPILATION;
  }
}

switch (compilationActivityMode) {
 ...
  default : CompilerDirectives.shouldNotReachHere("invalid activity mode")
}

```

Reply:
Done